chia/util/write_file.py

```diff
@@ -0,0 +1,24 @@
+import os
+import shutil
+
+from pathlib import Path
+from typing import Union
+
+
+def write_file(file_path: Path, data: Union[str, bytes], *, file_mode: int = 0o600):
+    """
+    Writes the provided data to a temporary file and then moves it to the final destination.
+    """
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(os.path.dirname(file_path), mode=dir_perms, exist_ok=True)
+
+    temp_path: Path = file_path.with_suffix("." + str(os.getpid()))
+    mode: str = "w" if type(data) == str else "wb"
+    with open(os.open(str(temp_path), os.O_CREAT | os.O_EXCL | os.O_WRONLY, file_mode), mode) as f:
+        f.write(data)
+    try:
+        os.replace(str(temp_path), file_path)
+    except PermissionError:
+        shutil.move(str(temp_path), str(file_path))
```

Comment:
if this call fails, we don't delete the temporary file. should we?

Reply:
I'm pretty sure that both this and `os.replace()` will fail on windows if some other process (or our process) has the file open. Perhaps it would be reasonable to retry a few times, if you make `write_file()` async, you could async sleep for 0.5 seconds or so and retry a few times, before failing.

Reply:
There were some edge cases that caused issues on Windows, as you hinted at. I've added retry logic, and updated the code such that the temp file is removed if the move fails.

Several tests to handle os.replace failures. The tests pass for me on macOS and Windows.